src/wmtk/PolygonMesh.cpp

```diff
@@ -0,0 +1,573 @@
+    if (e_flag_accessor.index_access().scalar_attribute(eid) == 0) {
+        return false;
+    }
+
+    // opp and the edge to halfedge and halfedge to edge maps are implicit
+    // We simply check that the corresponding halfedge ids 2e and 2e + 1 are in range
+    if (2 * eid + 1 >= capacity(PrimitiveType::HalfEdge)) {
+        return false;
+    }
+
+    return true;
+}
+
+bool PolygonMesh::is_face_connectivity_valid(long fid) const
+{
+    // Face id is in range
+    if ((fid < 0) || (fid >= capacity(PrimitiveType::Face))) {
+        return false;
+    }
+
+    // Face is not deleted
+    ConstAccessor<char> f_flag_accessor = get_flag_accessor(PrimitiveType::Face);
+    if (f_flag_accessor.index_access().scalar_attribute(fid) == 0) {
+        return false;
+    }
+
+    // The next orbit of the face halfedge has constant face index fid
+    ConstAccessor<long> fh_accessor = create_const_accessor<long>(m_fh_handle);
+    ConstAccessor<long> hf_accessor = create_const_accessor<long>(m_hf_handle);
+    ConstAccessor<long> next_accessor = create_const_accessor<long>(m_next_handle);
+    long hid_start = fh_accessor.index_access().scalar_attribute(fid);
+    long hid_iter = hid_start;
+    do {
+        if (hf_accessor.index_access().scalar_attribute(hid_iter) != fid) {
+            return false;
+        }
+        hid_iter = next_accessor.index_access().scalar_attribute(hid_iter);
+    } while (hid_iter != hid_start);
+
+    return true;
+}
+
+bool PolygonMesh::is_halfedge_connectivity_valid(long hid) const
+{
+    // Halfedge id is in range
+    long h_cap = capacity(PrimitiveType::HalfEdge);
+    if ((hid < 0) || (hid >= h_cap)) {
+        wmtk::logger().error("halfedge {} is out of range [0, {})", hid, h_cap);
+        return false;
+    }
+
+    // Halfedge is not deleted
+    ConstAccessor<char> h_flag_accessor = get_flag_accessor(PrimitiveType::HalfEdge);
+    if (h_flag_accessor.index_access().scalar_attribute(hid) == 0) {
+        wmtk::logger().error("halfedge {} is deleted", hid);
+        return false;
+    }
+
+    // next and prev are (locally) inverse
+    ConstAccessor<long> next_accessor = create_const_accessor<long>(m_next_handle);
+    ConstAccessor<long> prev_accessor = create_const_accessor<long>(m_prev_handle);
+    long hid_next = next_accessor.index_access().scalar_attribute(hid);
+    long hid_np = prev_accessor.index_access().scalar_attribute(hid_next);
+    if (hid_np != hid) {
+        wmtk::logger().error("halfedge {} is next of {} but has prev {}", hid_next, hid, hid_np);
+        return false;
+    }
+    long hid_prev = prev_accessor.index_access().scalar_attribute(hid);
+    long hid_pn = next_accessor.index_access().scalar_attribute(hid_prev);
+    if (hid_pn != hid) {
+        wmtk::logger().error("halfedge {} is prev of {} but has next {}", hid_prev, hid, hid_pn);
+        return false;
+    }
+
+    return true;
+}
+
+long PolygonMesh::count_vertices_slow() const
```

Comment:
this and the next function's names don't clearly indicate their purpose - i think there's an abstract interest in this count function and perhaps a purpose for debugging, but the real concept is to generate vertices and faces from halfedges which "weirdly"? doesn't quite belong in this class as initialization utilities should handle that. you could use a variation of this implementation to implement `out` / specify the number of vertices in teh system though? same with the count_faces_slow (in the sense you could use the PolygonMesh class to compute the number of vertices + faces and construct the pertinent maps)

Reply:
(Discussed in person)